internal/client/client.go

```diff
@@ -682,6 +682,34 @@ func SyncTimelineHasEventID(roomID string, eventID string) SyncCheckOpt {
 	})
 }
 
+// Check that the state for `roomID` has an event which passes the check function.
+func SyncStateHas(roomID string, check func(gjson.Result) bool) SyncCheckOpt {
+	return func(clientUserID string, topLevelSyncJSON gjson.Result) error {
+		err := loopArray(
+			topLevelSyncJSON, "rooms.join."+GjsonEscape(roomID)+".state.events", check,
+		)
+		if err == nil {
+			return nil
+		}
+		return fmt.Errorf("SyncStateHas(%s): %s", roomID, err)
+	}
+}
+
+// Check that the state for `roomID` has an event which matches the event ID.
+func SyncStateHasEventID(roomID string, eventID string) SyncCheckOpt {
```

Comment:
Please remove: this doesn't compose well at all. What if we wanted to check the timeline for an event ID? What about ephemeral data? Room account data? Then we'd end up with:
 - `SyncStateHasEventID`
 - `SyncTimelineHasEventID`
 - `SyncEphemeralHasEventID`
 - `SyncRoomAccountDataHasEventID`

I hope it's clear how unruly and unmanageable this ends up becoming, especially if you then want to check the type/state key like `SyncStateHasStateKey` does. It would be better if you added functions which match `func(gjson.Result) bool` and compose well e.g:
 - `func CheckEventIDIs(eventID) func(gjson.Result) bool`
 - `func CheckEventTypeIs(eventType) func(gjson.Result) bool`

This would end up looking something like:
```go
SyncStateHas(roomID, CheckEventIDIs(eventID), CheckEventTypeIs(eventType))
```

But this is also not great because often you want to check a few fields, though which fields depends on the test. Also, what is the relationship between the check functions: are they ANDed or ORd? At this point, it is often easier just to write the check function yourself:
```go
SyncStateHas(roomID, func(ev gjson.Result) bool {
    		return ev.Get("type").Str == eventType && ev.Get("state_key").Str == stateKey
})
```

which also allows the flexibility to check say the `content` field easily enough. This is why we, as of yet, do not have per-event  check functions: the benefits are unclear and the assertion itself is unclear.

Reply:
Do note that:
* The events in the `ephemeral` and `account_data` sections do not have `event_id`s and we would not want to define `SyncEphemeralHasEventID` and `SyncRoomAccountDataHasEventID` as the spec stands.
* `SyncTimelineHasEventID` already exists in Complement and this PR follows the existing pattern.


Shall we remove `SyncTimelineHasEventID` then?

Reply:
Yes